### PR TITLE
Initial commit of revised docs

### PR DESCRIPTION
**Pull request by rowanwins**
Gday @DenisCarriere & @stebogit 

Here is an initial commit of the new docs. 

If you pull them down you should be able to run `npm run serveStatic` to see them in action.

A few things to note
- `/scripts/parseJSON` takes the JSON output from documentation.js and translates it into something more usable by the app. This is probably the trickiest part of everything I've done - trying to parse the original JSON output from documentation.js. Im thinking I might be able to change this workflow a bit to use documentation.js itself to parse the JSON using although I need to play around to better understand how these functions work
````
require('documentation').util.createFormatters,
require('documentation').util.createLinkerStack,
````
- because of the complexity of the above there are a few anomolies in the presentation of things
- I still need to pull down the latest version of the turf repo to include the new options approach but i'll get on to that now

All in all Im really happy with how they look and I think they're a good step forward.

### CODE REVIEW

**Comment by DenisCarriere** [comment]
> I still need to pull down the latest version of the turf repo to include the new options approach but i'll get on to that now

What if we create a documentation JSON object directly in Turf and that way `turf-www` would use that to build the docs. That we we don't have to go back and forth between Turf & Turf-www to compile the JSON documentation object.

>  translates it into something more usable by the app. This is probably the trickiest part of everything I've done - trying to parse the original JSON output from documentation.js

We can definitely try to help out with that part, the documentation.js JSON object could definitely be simplified for a front end app.

File: scripts/parseJson.js [hunk 1]
```diff
@@ -0,0 +1,171 @@
+const loadJsonFile = require('load-json-file');
+const writeJsonFile = require('write-json-file');
+
+loadJsonFile('./scripts/tempConfig.json').then(json => {
```
Comment (DenisCarriere) [inline scripts/parseJson.js]: I'd recommend you use `path` to define file paths, that way it's compatible with Mac & Windows also use `__dirname` to define the absolute filepath that way you can execute the script from any directory.

```js
path.join(__dirname, 'tempConfig.json')
```

File: scripts/parseJson.js [hunk 2]
```diff
@@ -0,0 +1,171 @@
...
+	var moduleSidebarList = []
+	var completeModules = []
+
+    json.forEach(function (module) {
```
Comment (DenisCarriere) [inline scripts/parseJson.js]: You can use `load.sync` and chain this `forEach` into a simple to understand line.

```js
var moduleSidebarList = []
var completeModules = []
load.sync(path.join(__dirname, 'tempConfig.json')).forEach(module => {
```

File: scripts/parseJson.js [hunk 3]
```diff
@@ -0,0 +1,171 @@
...
+      moduleSidebarList.push({
+        isHeading: isHeading,
+        name: module.name,
+        hidden: false
+      })
+      if (!isHeading) {
+        var parent = getParent(module.context.file)
+	      completeModules.push({
+	      	name: module.name,
+          description: concatTags(module.description.children[0].children, true),
```
Comment (DenisCarriere) [inline scripts/parseJson.js]: Seems like you would need to catch if there's `description` is defined.

```
/Users/mac/Github/turf-www/scripts/parseJson.js:69
      desc: concatTags(result.description.children[0].children, false)
                                                     ^

TypeError: Cannot read property 'children' of undefined
```

File: scripts/parseJson.js [hunk 4]
```diff
@@ -0,0 +1,171 @@
...
+	      	throws: getThrows(module.throws)
+	      })
+      }
+	
+	})
+	var config = {
+    	sidebar: moduleSidebarList,
+    	modules: completeModules
+    }
+  writeJsonFile('./template/turf-template/src/assets/config.json', config).then(() => {
```
Comment (DenisCarriere) [inline scripts/parseJson.js]: Use the `sync` method to make sure your next lines of code are executed in proper order.

Promises are good when doing HTTP requests, but handling local file system stuff I'd simply just use `sync` that way you code is executed line by line.

```js
write.sync(path.join(__dirname, '..', 'template', 'turf-template', 'src', 'assets', 'config.json'), config)
console.log('Done compiling the config file')
```

**Comment by DenisCarriere** [comment]
😨 So much webpack configs...

File: template/turf-template/package.json [hunk 5]
```diff
@@ -0,0 +1,74 @@
+{
+  "name": "turf-template",
+  "version": "1.0.0",
+  "description": "Turf.js documentation",
+  "author": "Rowan Winsemius <rowanwins@yahoo.com.au>",
```
Comment (DenisCarriere) [inline template/turf-template/package.json]: Do you want robots sending you junk mail?

File: template/turf-template/package.json [hunk 6]
```diff
@@ -0,0 +1,74 @@
+{
+  "name": "turf-template",
```
Comment (DenisCarriere) [inline template/turf-template/package.json]: Why can't this `package.json` be merged into the root directory?

File: package.json [hunk 7]
```diff
@@ -28,16 +28,30 @@
...
     "load-json-file": "^3.0.0",
+    "optimize-css-assets-webpack-plugin": "^3.2.0",
+    "ora": "^1.3.0",
+    "rimraf": "^2.6.2",
+    "semver": "^5.4.1",
     "serve": "^6.2.0",
+    "shelljs": "^0.7.8",
+    "standard": "^10.0.3",
     "static-server": "^2.0.3",
+    "webpack": "^3.7.1",
```
Comment (DenisCarriere) [inline package.json]: @rowanwins Ops, didn't notice the 2nd `package.json` in `turf-template`, we shouldn't have two `packages.json` files

File: scripts/parseJson.js [hunk 8]
```diff
@@ -0,0 +1,168 @@
...
+  moduleSidebarList.push({
+    isHeading: isHeading,
+    name: module.name,
+    hidden: false
+  })
+  if (!isHeading) {
+    var parent = getParent(module.context.file)
+    completeModules.push({
+      name: module.name,
+      description: concatTags(module.description.children[0].children, true),
```
Comment (DenisCarriere) [inline scripts/parseJson.js]: This one is very problematic, might be worth doing a `getDescription` function to handle this.

File: scripts/parseJson.js [hunk 9]
```diff
@@ -0,0 +1,168 @@
+const path = require('path')
+const load = require('load-json-file')
+const write = require('write-json-file')
+
+var moduleSidebarList = []
+var completeModules = []
+load.sync(path.join(__dirname, 'tempConfig.json')).forEach(module => {
```
Comment (DenisCarriere) [inline scripts/parseJson.js]: `module` is a reserved name, could be changed to `metadata` instead

**Comment by DenisCarriere** [comment]
@rowanwins SUPER cool! 🎉 

turf

**Comment by DenisCarriere** [comment]
@rowanwins don't feel bad for my next commits, this setup is WAYYYYY too complicated for what it's worth.

It's only static HTML, no need for crazy webpack production/dev/hot reloading.

Making it the setup too difficult to understand will deter people from contributing.

**Comment by DenisCarriere** [comment]
@rowanwins We should add an "expand" button for the map, that way the users can see it full screen.

image

**Comment by DenisCarriere** [comment]
We should include Javascript syntax highlighting for the `@example` section.

image

**Comment by DenisCarriere** [comment]
Handling of `@callback` JSDocs might be needed in the `config.json` to properly handle that.

image

**Comment by DenisCarriere** [comment]
@rowanwins Dropped the "Table of Contents" from `documentation.yml` since we always forget to add the modules to that config.

It would be best to use JSDocs to organize the modules sidebar.

We use `@since` to organize when the module was released.

http://usejsdoc.org/tags-since.html

And a `@category` tag to group them into sections. (not native to JSDocs)

https://github.com/ErnstHaagsman/jsdoc-plugins/blob/master/categories.md

**Comment by DenisCarriere** [comment]
@rowanwins I should be done for the night 👍 ... lol! You'll ask where all your code went! But it all still works.

How to run it

```bash
$ git clone git@github.com:Turfjs/turf-www.git --recursive
$ cd turf-www
$ yarn
$ npm run build
$ npm run start
```

Done! 🎉 Hot reloading is enabled, so you can edit your Vue pages without having to re-create the webpack bundle.

File: template/turf-template/package.json (hunk 5 above)
Comment (rowanwins) [reply template/turf-template/package.json]: yes please - I love it!

**Comment by rowanwins** [comment]
Gday @DenisCarriere 

All good, I'm used to your 'tweaking' :) 

- I was just running with the default [vue webpack template](https://github.com/vuejs-templates/webpack) which provides more settings than what we need.
- Re using a `category` tag sounds sensible instead of the yml file
- Highlighting would be good, I did experiment with this but couldn't find any colours that really made it worth it so i ditched it for the timebeing in the interests of time
- I also need to add a github button in somewhere
- I'll have a play with the parseJson scipt tonight to see if introducing documentation.js that offers us any benefits in terms of the handling all those variations eg callbacks, options.
  - There is so much involved in parsing that original json object, the old process hid all that complexity so well, had I known what was involved....
  - I'll also work on constructing the options into something nicely formatted.

**Comment by rowanwins** [comment]
And in case you didn't know Vue has an awesome [dev tools extension for chrome](https://chrome.google.com/webstore/detail/vuejs-devtools/nhdogjmejiglipccpnnnanhbledajbpd?hl=en)

**Comment by rowanwins** [comment]
Gday @DenisCarriere 

So I've added support for options now, I think it's working reasonably well.

The thing I've noticed is that the modules that are parents (eg helpers) don't seem to work properly. For example `earthRadius` is currently being published from `@turf/helpers`, so we obviously need to tweak that `create-config` script a bit. 

A few of the maps are broken at the moment because we pull in turf via a cdn, so it's a bit hard to test the maps at the moment, but theoretically we could build a version of turf to use for testing relatively easily...

PS perhaps we could also hijack the 'units' (eg metres, km's, miles) property during the `create-config` script to have the same description, at the moment we've got all sorts of variations across the modules

**Comment by DenisCarriere** [comment]
@rowanwins Nice stuff! There's definitely room for improvement, it's heading in the right direction! We will continue hacking on this branch until it's ready enough to be merged.

File: template/turf-template/package.json (hunk 5 above)
Comment (DenisCarriere) [reply template/turf-template/package.json]: 🤣

**Comment by DenisCarriere** [comment]
@rowanwins I've got the syntax highlighting working, but it doesn't work perfectly (might be a Vue thing).

image

**Comment by rowanwins** [comment]
Gday @DenisCarriere 

Fixed the code highlighting stuff. I think it's because prism runs when the dom is loaded but doesnt neceassarily watch for changes to the dom.

I had a look into the category plugin tag for jsdoc but it looked a bit complicated in that it required a jsdoc config file (that we dont currently have) and it looks the like the documenation.js team weren't in favour of it (see [here](https://github.com/documentationjs/documentation/issues/100) and [here](https://github.com/documentationjs/documentation/issues/128) ). So for the timebeing I've reinstated the yaml file (which I don't actually mind that much).

So I think the main thing left is a good comb through to check for consistency. How is the main turf repo tracking, eg if I build it will I get all the expected stuff contained within v5?

**Comment by DenisCarriere** [comment]
🎉 Woot! The code highlighting works as intended, good job @rowanwins!

image

> So for the timebeing I've reinstated the yaml file (which I don't actually mind that much).

👍 Sounds good to me, I was thinking the same after pondering about it, the `@category` tag seems like a pain, especially if it's not native to JSDocs.

YAML config 👍 

> if I build it will I get all the expected stuff contained within v5?

Yep, you should get everything updated with the new `options` params.

Right now I'm trying to "clean" out some ES module stuff, any future new modules will 100% need to support ES modules or be built-in TurfJS modules, no more CommonJS coming into Turf.

**Comment by DenisCarriere** [comment]
@rowanwins Do you think we could add a Fullscreen button to the map, it's good that it's tiny.. but... hard to take print-screens with such a tiny map.

**Comment by rowanwins** [comment]
Will do. 

And how good is the layer control, I give it 10 out of 10 (although I'm biased!)

**Comment by DenisCarriere** [comment]
@rowanwins 2nd thing... Could we include links in the Arguments.

Type => `Feature<LineString>`

image

3rd... The links seem to be pointing to the outdated GeoJSON docs, the website is pretty broken at the moment.

http://geojson.org/geojson-spec.html#linestring

image

**Comment by DenisCarriere** [comment]
haha!! I noticed that layer control! 🏅 10/10 Not biased whatsoever!

**Comment by DenisCarriere** [comment]
For those who haven't seen it:

## New Layer Control

image

**Comment by DenisCarriere** [comment]
@rowanwins http://localhost:8080/docs#random doesn't work?... I know I've reworked that one... maybe I just need to publish a new npm release for that one.

**Comment by rowanwins** [comment]
Hi @DenisCarriere 

Currently the turf maps are being generated using a cdn version of turf (presumably 4.something) so Im not surprised it's not working, this approach works kinda nicely for the site because `turf` is exposed as a global variable.

For the fullscreen map I'll just add this [leaflet plugin](http://leaflet.github.io/Leaflet.fullscreen/).

Yeah it's a shame about the geojson.org docs, not sure what's going on there... I can look at adding the links in to the arguments, I originally left them out just because it felt like link overkill for some modules that supported lots of types. Anyway I'll take another look and see if I can work out a way.

**Comment by DenisCarriere** [comment]
@rowanwins 👍 Nice stuff

Once we bump the CDN we should be fine, it's fine the way it is now. Just need to publish v5.0 (eventually... 😛 )

**Comment by rowanwins** [comment]
Could we publish it with the beta tag so it's not the default turfjs install but is still available? eg https://gist.github.com/coolaj86/1318304#beta-and-release-versions

**Comment by DenisCarriere** [comment]
@rowanwins we could... however I've always found those beta release tags as buggy, I dunno, we should be able to release something that works (that's the goal).

**Comment by DenisCarriere** [comment]
Fixed conflicts, this branch is ready to be merged.

@rowanwins Any last updates to modify? I'm good on my side

**Comment by rowanwins** [comment]
Oooo umm I wouldn't mind doing a check tonight pulling in a built version of v5, I should be able to do that. Just wanna go through all the modules top to bottom and make sure we dont get any unexpected errors. It certainly should be pretty close.

**Comment by rowanwins** [comment]
ps did you find my special effects?!

**Comment by DenisCarriere** [comment]
@rowanwins The sidebar transitions 👍 I saw the commit, but I didn't see what was changed.

**Comment by rowanwins** [comment]
Gday @DenisCarriere 

Just tried to build a new version of turf.js to do a test on the docs but got this
`````
cd turf/packages/turf
npm run pretest

index.js → turf.js...
[!] Error: Unexpected token
node_modules\@turf\nearest-point-to-line\node_modules\geodesy\latlon-spherical.js (35:0)
33:  *     var p1 = new LatLon(52.205, 0.119);
34:  */
35: function LatLon(lat, lon) {
    ^
36:     // allow instantiation without 'new'
37:     if (!(this instanceof LatLon)) return new LatLon(lat, lon);
`````
Any ideas? Been yonks since I've built turf myself!

**Comment by DenisCarriere** [comment]
@rowanwins Did you use `lerna bootstrap` to update all the dependencies?

I'm pretty sure you are using old `node_modules` since we dropped `geodesy`, that dependency is not internal.

**remove all node_modules and re-install**

```
$ rm -r packages/turf*/node_modules
$ npm t
```

**Comment by rowanwins** [comment]
Managed to get a build completed although it looks like the built files produced by rollup aren't quite right...

With the `turf.js` in the browser I get `turf is not defined`
With the `turf.min.js` in the browser I get errors like `turf.lineString is not a function`

Am I do something wrong or am I breaking new ground?

**Comment by DenisCarriere** [comment]
You are breaking new ground... 🚀 🌔 

I don't know yet how I'm going to be tackling the `@turf/turf` since `jsts` and most of the Rollup (ES Modules) dependencies are using ES6+ code and depend on running Babel to compile the code to ES5.

Today I'm going to attempt to publish all modules to `v5.0` and then afterwards tackle `@turf/turf` since that will require a bit more attention.

**Comment by DenisCarriere** [comment]
> With the turf.js in the browser I get turf is not defined

You need to use `UMD` for the browser bundle, right now it's configured for CommonJS which we should change for both UMD & UMD.

https://github.com/Turfjs/turf/blob/master/packages/turf/rollup.config.js

**Comment by DenisCarriere** [comment]
Going to merge this at the same time as the v5.0 release

**Comment by DenisCarriere** [comment]
@rowanwins Just published v5.0 so hopefully these docs won't break 🤞

**Pull request merged**